### PR TITLE
feat: wire snapshot_and_clear API to swap logic

### LINKED ISSUE #47: [W10d] infmon_snapshot_and_clear API message
## Spec
[004-backend-architecture](https://github.com/r12f/InFMon/blob/main/specs/004-backend-architecture.md)

## Scope
- Wire the infmon_snapshot_and_clear(flow_rule_id) API message to the swap logic from W8
- Return the descriptor of the retired table (generation G)
- Ensure the reply contains all fields needed by the frontend to walk the retired table

## Dependencies
Blocked by: #44, #42

### PR DESCRIPTION
**Pull request by r12f**
## Summary

Wire `infmon_snapshot_and_clear(flow_rule_id)` API message to the swap logic from W8.

### Changes

- **api_handler.h**: Add `snap_mgr` and `flow_rule_ids[]` to `infmon_api_ctx_t`, new error codes (`ALLOC_FAILED`, `TOO_MANY_RETIRED`, `NULL_TABLE`, `NO_SNAPSHOT_MGR`), `infmon_api_snap_reply_t` struct, and declarations for `snapshot_and_clear` and `flow_rule_add_with_id`
- **api_handler.c**: Implement `infmon_api_snapshot_and_clear()` — resolves flow_rule_id → index, calls snapshot manager, publishes stats descriptor for the retired table; add `find_rule_index_by_id()` helper, `map_snap_result()`, `flow_rule_add_with_id()`, and ID compaction on delete
- **CMakeLists.txt**: Link `infmon_snapshot` to api_handler and test targets
- **test_api_handler.cc**: 7 new tests covering basic swap, not-found ID, null ctx, missing snap_mgr, descriptor fields, multiple swaps, and ID preservation across delete/compaction

### Test Plan

All 8 test suites pass (23 api_handler tests total, 7 new).

Closes #47

### CODE REVIEW

File: src/backend/src/api_handler.c [hunk 1]
```diff
@@ -106,9 +144,26 @@ infmon_api_result_t infmon_api_flow_rule_add(infmon_api_ctx_t *ctx, const infmon
     }
     ctx->tables[idx] = ct;
 
+    /* 5. Record the flow_rule_id if provided. */
+    if (id) {
+        ctx->flow_rule_ids[idx] = *id;
```
Comment (banidoru) [inline src/backend/src/api_handler.c:150]: When `id` is NULL (i.e. the plain `infmon_api_flow_rule_add` path), `ctx->flow_rule_ids[idx]` is left with whatever was there before. If the caller later adds a rule via `add_with_id`, then deletes it, then adds another via plain `add` — the slot could retain a stale ID that `find_rule_index_by_id` would match.

Consider adding an `else { memset(&ctx->flow_rule_ids[idx], 0, sizeof(ctx->flow_rule_ids[0])); }` branch.

File: src/backend/src/api_handler.c [hunk 2]
```diff
@@ -183,3 +241,58 @@ infmon_api_result_t infmon_api_flow_rule_get_by_name(const infmon_api_ctx_t *ctx
     }
     return INFMON_API_ERR_NOT_FOUND;
 }
+
+/* ── Snapshot and clear ──────────────────────────────────────────── */
+
```
Comment (banidoru) [inline src/backend/src/api_handler.c:249]: No null-check on `reply` before `memset(reply, 0, ...)`. If a caller passes NULL this will segfault. Either add a guard or document that `reply` must be non-NULL in the header.

File: src/backend/src/api_handler.c [hunk 3]
```diff
@@ -183,3 +241,58 @@ infmon_api_result_t infmon_api_flow_rule_get_by_name(const infmon_api_ctx_t *ctx
...
+        return;
+    }
+
+    /* Resolve flow_rule_id → index. */
+    uint32_t idx = find_rule_index_by_id(ctx, flow_rule_id);
+    if (idx == (uint32_t) -1) {
+        reply->result = INFMON_API_ERR_NOT_FOUND;
+        return;
+    }
+
```
Comment (banidoru) [inline src/backend/src/api_handler.c:281]: `infmon_flow_rule_get(ctx->rule_set, idx)` could theoretically return NULL if the rule set is in an inconsistent state. Dereferencing `->key_width` without a null-check is risky. A defensive check here would prevent a crash in edge cases.

File: src/backend/src/api_handler.c [hunk 4]
```diff
@@ -183,3 +241,58 @@ infmon_api_result_t infmon_api_flow_rule_get_by_name(const infmon_api_ctx_t *ctx
...
+                             infmon_flow_rule_get(ctx->rule_set, idx)->key_width, &snap_reply);
+
+    reply->result = map_snap_result(snap_reply.result);
+
+    if (snap_reply.result != INFMON_SNAP_OK)
+        return;
+
+    /* Build the descriptor from the retired table. */
+    infmon_counter_table_t *retired = snap_reply.retired_table;
+    infmon_stats_descriptor_t *desc = &reply->descriptor;
```
Comment (banidoru) [inline src/backend/src/api_handler.c:306]: When `ctx->stats_reg` is NULL, `seg_base` is 0 and the offsets become raw pointer addresses. Is this intentional? If `stats_reg` is required, it should be validated at the top of the function alongside `snap_mgr`. If optional, the descriptor offsets are meaningless without a valid base.

**Review by banidoru — COMMENTED** [review]
- `flow_rule_add_internal`: when `id` is NULL (plain `add` path), `flow_rule_ids[idx]` is not zeroed — stale data could cause `find_rule_index_by_id` to match a wrong rule
- `infmon_api_snapshot_and_clear`: no null-check on `reply` parameter before `memset`
- `infmon_api_snapshot_and_clear`: `infmon_flow_rule_get()` return not null-checked before dereferencing `->key_width`
- Tests are thorough and cover the key scenarios well; good compaction test

File: src/backend/include/infmon/api_handler.h [hunk 5]
```diff
@@ -100,6 +128,19 @@ infmon_api_result_t infmon_api_flow_rule_get_by_name(const infmon_api_ctx_t *ctx
...
+ *
+ * On success the reply's descriptor is fully populated with the
+ * retired table's metadata (generation, offsets, counters).
+ *
+ * Requires ctx->snap_mgr to be set (returns INFMON_API_ERR_NO_SNAPSHOT_MGR
+ * otherwise).
+ */
+void infmon_api_snapshot_and_clear(infmon_api_ctx_t *ctx, infmon_flow_rule_id_t flow_rule_id,
+                                   infmon_api_snap_reply_t *reply);
+
```
Comment (banidoru) [inline src/backend/include/infmon/api_handler.h:148]: `infmon_api_snapshot_and_clear` returns `void`, unlike every other function in this API which returns `infmon_api_result_t`. This forces callers to use a different error-checking pattern just for this one call. Consider returning `infmon_api_result_t` directly (the reply struct can still carry the descriptor) for API consistency.

File: src/backend/src/api_handler.c [hunk 6]
```diff
@@ -183,3 +241,58 @@ infmon_api_result_t infmon_api_flow_rule_get_by_name(const infmon_api_ctx_t *ctx
...
+
+/* ── Snapshot and clear ──────────────────────────────────────────── */
+
+void infmon_api_snapshot_and_clear(infmon_api_ctx_t *ctx, infmon_flow_rule_id_t flow_rule_id,
+                                   infmon_api_snap_reply_t *reply)
+{
+    memset(reply, 0, sizeof(*reply));
+
+    if (!ctx) {
+        reply->result = INFMON_API_ERR_INVALID_RULE;
```
Comment (banidoru) [inline src/backend/src/api_handler.c]: Using `INFMON_API_ERR_INVALID_RULE` for a null `ctx` is misleading — the problem isn't an invalid rule, it's a null context. Consider adding `INFMON_API_ERR_INVALID_CTX` or reusing `INFMON_API_ERR_INTERNAL`.

File: src/backend/test/test_api_handler.cc [hunk 7]
```diff
@@ -244,3 +246,155 @@ TEST_F(ApiHandlerTest, ListAfterDeleteReflectsRemoval)
...
+  protected:
+    void SetUp() override
+    {
+        rule_set_ = infmon_flow_rule_set_create(INFMON_FLOW_RULE_MAX_KEYS_BUDGET);
+        ASSERT_NE(rule_set_, nullptr);
+        infmon_stats_registry_init(&stats_reg_, /* segment_base */ 0);
+
+        snap_mgr_ = new infmon_snapshot_mgr_t;
+        infmon_snapshot_mgr_init(snap_mgr_, /* num_workers */ 2, /* grace_ns */ 0, fake_clock_ns);
+
```
Comment (banidoru) [inline src/backend/test/test_api_handler.cc:269]: Allocating `infmon_snapshot_mgr_t` with C++ `new`/`delete` is fragile — if this struct ever gets a C-side allocation contract (e.g., `infmon_snapshot_mgr_create()` returning a malloc'd pointer), this will silently diverge. Consider using `malloc`/`free` or a stack allocation to stay closer to the C API conventions.

File: src/backend/test/test_api_handler.cc [hunk 8]
```diff
@@ -244,3 +246,155 @@ TEST_F(ApiHandlerTest, ListAfterDeleteReflectsRemoval)
...
+    /* Delete first rule — rule_bb should compact to index 0 with its ID intact. */
+    EXPECT_EQ(infmon_api_flow_rule_del(&ctx_, "rule_aa"), INFMON_API_OK);
+
+    infmon_api_snap_reply_t reply;
+    infmon_api_snapshot_and_clear(&ctx_, id_bb, &reply);
+    ASSERT_EQ(reply.result, INFMON_API_OK);
+    EXPECT_EQ(reply.descriptor.flow_rule_id.hi, 0x30u);
+    EXPECT_EQ(reply.descriptor.flow_rule_id.lo, 0x40u);
+    EXPECT_EQ(reply.descriptor.flow_rule_index, 0u);
+}
```
Comment (banidoru) [inline src/backend/test/test_api_handler.cc:400]: There's no test for calling `snapshot_and_clear` on a rule added via the plain `infmon_api_flow_rule_add` (without an ID). This is the path most likely to hit a stale-ID or zero-ID matching bug. Consider adding a test that adds a rule without an ID, then verifies `snapshot_and_clear` with a zero ID either works correctly or returns NOT_FOUND.

**Review by banidoru — COMMENTED** [review]
- `snapshot_and_clear` returns `void` while every other API function returns `infmon_api_result_t` — inconsistent interface; callers must know to dig into the reply struct instead of checking the return value.
- Null ctx maps to `INFMON_API_ERR_INVALID_RULE` which is semantically wrong — this is a null context, not an invalid rule.
- No test coverage for the interaction between plain `flow_rule_add` (no ID) and `snapshot_and_clear` — this is the path most likely to hit the stale-ID bug.
- Test uses C++ `new`/`delete` for a C struct (`infmon_snapshot_mgr_t`) — works but fragile if the struct ever gains a non-trivial C initializer.

File: src/backend/src/api_handler.c (hunk 1 above)
Comment (r12f) [reply src/backend/src/api_handler.c:150]: Good catch — when uid=1001(r12f) gid=1002(r12f) groups=1002(r12f),27(sudo) is NULL the slot was indeed left with stale data. Added an  branch that zeroes it out with . Fixed in 2e7b6f5.

File: src/backend/src/api_handler.c (hunk 2 above)
Comment (r12f) [reply src/backend/src/api_handler.c:249]: Agreed. Added a  guard at the top of the function, and documented the non-NULL requirement in the header. Fixed in 2e7b6f5.

File: src/backend/src/api_handler.c (hunk 3 above)
Comment (r12f) [reply src/backend/src/api_handler.c:281]: Valid point. Extracted the  call into a local variable with a NULL check that returns  before dereferencing. Fixed in 2e7b6f5.

File: src/backend/src/api_handler.c (hunk 4 above)
Comment (r12f) [reply src/backend/src/api_handler.c:306]: You're right —  should be required here since offsets are meaningless without a valid segment base. Changed the ternary to an explicit NULL check that returns  early, and updated the header doc to note  must be set. Fixed in 2e7b6f5.

File: src/backend/src/api_handler.c [hunk 9]
```diff
@@ -45,6 +64,20 @@ static uint32_t find_rule_index(const infmon_flow_rule_set_t *set, const char *n
     return (uint32_t) -1;
 }
 
+/**
+ * Find the index of a rule by flow_rule_id in the context.
+ * Returns the index, or (uint32_t) -1 if not found.
+ */
+static uint32_t find_rule_index_by_id(const infmon_api_ctx_t *ctx, infmon_flow_rule_id_t id)
+{
+    uint32_t n = infmon_flow_rule_count(ctx->rule_set);
```
Comment (banidoru) [inline src/backend/src/api_handler.c:73]: `find_rule_index_by_id` will match on `{0, 0}` — which is the default for any slot added via the plain `infmon_api_flow_rule_add` path (since `flow_rule_ids[]` isn't explicitly zeroed on add, see related comment). If two rules are added without IDs, calling `snapshot_and_clear` with a zero ID silently picks the first one. Consider rejecting zero IDs here as a sentinel, or documenting that `snapshot_and_clear` is only valid for rules added via `add_with_id`.

File: src/backend/src/api_handler.c [hunk 10]
```diff
@@ -183,3 +244,71 @@ infmon_api_result_t infmon_api_flow_rule_get_by_name(const infmon_api_ctx_t *ctx
...
+        return;
+    }
+    infmon_snapshot_and_clear(ctx->snap_mgr, ctx->tables, idx, INFMON_FLOW_RULE_SET_MAX,
+                              rule->key_width, &snap_reply);
+
+    reply->result = map_snap_result(snap_reply.result);
+
+    if (snap_reply.result != INFMON_SNAP_OK)
+        return;
+
```
Comment (banidoru) [inline src/backend/src/api_handler.c:303]: The descriptor captures offsets into `snap_reply.retired_table`'s memory, but nothing here retains ownership or prevents the snapshot manager from freeing the retired table before the caller consumes the descriptor. Document the lifetime contract — does the caller need to call a release function? Does `grace_ns` in the snapshot manager guarantee a window?

**Review by banidoru — COMMENTED** [review]
- `flow_rule_ids[]` is never zeroed in `infmon_api_ctx_init` — garbage IDs could cause `find_rule_index_by_id` to false-match before any rule is added; add a `memset` in init
- Two rules added via plain `infmon_api_flow_rule_add` both get zero IDs, so `find_rule_index_by_id({0,0})` silently returns the first — either document that snapshot requires `add_with_id`, or reject zero IDs in `find_rule_index_by_id`
- Retired table lifetime unclear: descriptor holds offsets into the retired table's memory, but nothing in this API documents or enforces that the retired table stays alive long enough for the caller to read it
- No concurrency guards — `snapshot_and_clear` does a table pointer swap that races with any concurrent reader of `ctx->tables[idx]`

File: src/backend/include/infmon/api_handler.h (hunk 5 above)
Comment (r12f) [reply src/backend/include/infmon/api_handler.h:148]: Good point — changed the return type to `infmon_api_result_t` for API consistency. The same error code is now returned directly AND stored in `reply->result` so callers can use either pattern. Updated the header doc accordingly. Fixed in f7761a5.

File: src/backend/src/api_handler.c (hunk 6 above)
Comment (r12f) [reply src/backend/src/api_handler.c]: Agreed, `ERR_INVALID_RULE` was misleading for a null context. Changed to `INFMON_API_ERR_INTERNAL` which better reflects a programming error. Adding a dedicated `ERR_INVALID_CTX` could be a follow-up if more context-validation cases arise. Fixed in f7761a5.

File: src/backend/test/test_api_handler.cc (hunk 7 above)
Comment (r12f) [reply src/backend/test/test_api_handler.cc:269]: Fair point about potential divergence. However, this is test-only code and the struct is currently a plain C struct with no special allocator. Switching to malloc/free here would add noise without improving correctness today. If/when `infmon_snapshot_mgr_create()` lands, we should update the test fixture to use it — but that's a separate concern. Leaving as-is for now.

File: src/backend/test/test_api_handler.cc (hunk 8 above)
Comment (r12f) [reply src/backend/test/test_api_handler.cc:400]: Added `SnapshotAndClearRejectsZeroId` test: adds a rule via plain `infmon_api_flow_rule_add` (no ID), then verifies that `snapshot_and_clear` with a zero ID returns `NOT_FOUND` rather than silently matching. Fixed in f7761a5.

File: src/backend/src/api_handler.c (hunk 9 above)
Comment (r12f) [reply src/backend/src/api_handler.c:73]: Great catch. Added an early-return guard: `infmon_flow_rule_id_is_zero(flow_rule_id)` now returns `INFMON_API_ERR_NOT_FOUND` before the search loop, treating zero as a sentinel for "no ID assigned". Combined with the earlier fix that zeroes out slots on plain add, this prevents any ambiguous zero-ID matching. Documented in the header. Fixed in f7761a5.

File: src/backend/src/api_handler.c (hunk 10 above)
Comment (r12f) [reply src/backend/src/api_handler.c:303]: The lifetime contract is governed by the snapshot manager's `grace_ns` parameter: retired tables are held for at least `grace_ns` nanoseconds before being recycled. In this PR's test fixture `grace_ns` is 0 (immediate reuse is safe because tests consume the descriptor synchronously). In production, `grace_ns` is set by the daemon to provide a consumption window. I've added a note to the header doc clarifying that `snap_mgr` and `stats_reg` must be set. A dedicated `infmon_api_snap_release()` or explicit lifetime doc could be a follow-up — the current contract is inherited from the snapshot manager layer.

File: src/backend/src/api_handler.c [hunk 11]
```diff
@@ -183,3 +244,80 @@ infmon_api_result_t infmon_api_flow_rule_get_by_name(const infmon_api_ctx_t *ctx
...
+    /* Build the descriptor from the retired table. */
+    infmon_counter_table_t *retired = snap_reply.retired_table;
+    infmon_stats_descriptor_t *desc = &reply->descriptor;
+
+    desc->flow_rule_id = flow_rule_id;
+    desc->flow_rule_index = idx;
+    desc->generation = snap_reply.retired_generation;
+    desc->epoch_ns = retired->epoch_ns;
+
```
Comment (banidoru) [inline src/backend/src/api_handler.c:312]: Nit: `stats_reg` NULL check is placed after `infmon_snapshot_and_clear()` has already executed the swap. If `stats_reg` is NULL, the table has been swapped (retired table moved to snapshot manager) but the caller receives an error with no descriptor — leaving the system in a silently inconsistent state. Consider moving this check to the top of the function alongside the `snap_mgr` guard, so the swap never happens if the prerequisites aren't met.

**Review by banidoru — COMMENTED** [review]
Re-review (iteration 2) — all prior concerns addressed. Clean.

- All 10 prior review threads resolved: stale-ID zeroing, null-reply guard, null-rule check, stats_reg validation, return type consistency, error codes, zero-ID sentinel, test coverage.
- New `infmon_flow_rule_id_is_zero()` guard and `else { memset }` branch correctly prevent ambiguous zero-ID matching.
- `stats_reg` NULL check in `snapshot_and_clear` happens *after* the swap — if `stats_reg` is unexpectedly NULL, the table swap has already committed but the caller gets an error with no descriptor. Consider validating `stats_reg` alongside `snap_mgr` at the top of the function (minor, since `stats_reg` should always be set in practice).
- Test coverage is thorough: 9 new snapshot tests covering happy path, error paths, compaction, zero-ID rejection, and return-value consistency.
- Code is clean, well-documented, and consistent with the existing API style.

File: src/backend/src/api_handler.c (hunk 11 above)
Comment (banidoru) [reply src/backend/src/api_handler.c:312]: Still open — the `stats_reg` NULL check at line 305 remains after `infmon_snapshot_and_clear()` has already executed the swap. If `stats_reg` is NULL, the table gets retired but the caller receives an error with no descriptor, leaving the system in an inconsistent state. Please move this check above the `infmon_snapshot_and_clear()` call alongside the `snap_mgr` guard.

**Review by banidoru — CHANGES_REQUESTED** [review]
- All prior review feedback addressed well — stale-ID zeroing, NULL guards, return type consistency, zero-ID sentinel, and test coverage all look solid.
- **Still open:** `stats_reg` NULL check in `infmon_api_snapshot_and_clear` is placed *after* the swap — if `stats_reg` is NULL the table is already retired but no descriptor is returned, leaving an inconsistent state. Move the check before the `infmon_snapshot_and_clear()` call.
- Test coverage is comprehensive: zero-ID rejection, multiple swaps, ID preservation across delete, null ctx/snap_mgr paths all exercised.
- API surface is clean and consistent after the `infmon_api_result_t` return type change.

**Review by banidoru — CHANGES_REQUESTED** [review]
- Previous feedback mostly addressed — stale ID zeroing, null guards, return type consistency, zero-ID sentinel, and test coverage all look good.
- **Still open:** `stats_reg` NULL check remains after the swap call. If `stats_reg` is NULL, the table gets retired but the caller gets an error with no descriptor — inconsistent state. Must move this check before `infmon_snapshot_and_clear()`.
- No new issues found in the iteration 2 diff beyond the outstanding `stats_reg` ordering problem.
- Test suite is comprehensive — good coverage of edge cases (zero ID, null ctx, multiple swaps, delete-compact).

File: src/backend/src/api_handler.c (hunk 11 above)
Comment (r12f) [reply src/backend/src/api_handler.c:312]: Good catch — moved the `stats_reg` NULL check above the `infmon_snapshot_and_clear()` call so the swap never executes if we can't build a valid descriptor. Fixed in d4a8d26.

**Review by banidoru — APPROVED** [review]
- All prior review concerns addressed across 3 iterations — stale ID zeroing, NULL guards, stats_reg pre-check, return type consistency, zero-ID sentinel, and test coverage gaps all fixed.
- Defensive checks are thorough: NULL reply, NULL ctx, NULL snap_mgr, NULL stats_reg, NULL rule, zero ID — all guarded before any state mutation.
- ID compaction on delete correctly mirrors table compaction with trailing slot cleanup.
- Test suite is comprehensive: 9 snapshot tests covering happy path, error paths, multi-swap generations, ID preservation across compaction, and zero-ID rejection.
- Code is clean and well-structured. LGTM.

**Review by banidoru — APPROVED** [review]
- All prior review concerns addressed: stale-ID zeroing, NULL guards, stats_reg pre-swap check, return type consistency, zero-ID sentinel rejection
- Defensive checks are thorough — NULL reply, NULL ctx, NULL snap_mgr, NULL stats_reg, NULL rule all handled before any mutation
- Test coverage is solid: basic, not-found, no-snap-mgr, null-ctx, descriptor fields, multiple swaps, delete compaction, zero-ID rejection, return-type consistency
- Code is clean, well-commented, and API-consistent
- No new issues found in this iteration

**Review by banidoru — APPROVED** [review]
- All prior feedback addressed: stale-ID zeroing, NULL guards, return type consistency, stats_reg pre-swap check
- Defensive coding is thorough — every error path sets both return value and reply->result
- Zero-ID sentinel rejection prevents ambiguous matching on plain-add rules
- Test coverage is comprehensive: basic, not-found, null-ctx, no-snap-mgr, zero-ID, multi-swap, delete-compaction
- Clean, well-structured diff — LGTM

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.
